content/blog/2025-03-06-path-profiling.md

```diff
@@ -0,0 +1,98 @@
+
+The paper includes other details on placing instrumentation code for fake edges, placing instrumentation code efficiently, storing counts of path sums in arrays vs. hash maps, and selecting registers for storing the active path sum, but the key contribution is the algorithm described above.
+
+### Analysis ###
+
+The algorithm presented in this paper is interesting on pure theoretical grounds. It is always informative to find a minimum of anything, so the minimal encoding presented is valuable as a pure mathematical object. As a practical matter, the encoding is perhaps less valuable.
+
+It turned out (in the evaluation section) that many programs have too many paths to store path counts in an array, so the profiler presented in the paper used a hash map anyway. While having small integers to represent paths is valuable in its own right, the greater benefit was in being able to store counts in an array. This is born out by the data in the paper, with programs that were small enough to use an array for path counts having noticably lower overheads than those that required a hash map.
+
+A similar argument can be made about reducing profiling overhead in general. While making programs run faster is almost always a good thing, this particular improvement did not cross the line into "viable-in-production" (for some definition of production). So for example, it is difficult to imagine a JIT compiler tolerating a 30% overhead for a running program at any stage of compilation. In this sense, the progress made in this paper towards low-overhead path profiling is perhaps better viewed as a step in the right direction than as a paradigm shift.
+
+Besides the path profiling algorithm, the paper's main argument is in favor of path profiling itself, arguing that it produces substantially better profiles than edge profiling.
+
+The paper successfully argues that path profiling produces better profiles. The implemented path profiler reliably provides frequencies for paths that are meaningfully longer than those produced by the edge profiler they compare against (and, of course, it should produce nearly identical information about edge frequencies). But the paper fails to take the final step and argue that the longer paths lead to better optimized programs. The authors don't do any experiments where the path and edge profiles are used to separately optimize the profiled programs. Such an experiment would have completed their argument that path profiling is worth the extra cost.
```

Comment:
Not sure what you mean by "longer" here… are you saying that edge profiling _can_ give you perfectly accurate results, as long as the paths are short? Or just that the universal inaccuracy of edge profiling gets worse as paths get longer? Maybe it would be good to be clearer/more quantitative here.